lib/elixir/lib/kernel/utils.ex

```diff
@@ -108,6 +111,19 @@ defmodule Kernel.Utils do
     {struct, enforce_keys, Module.get_attribute(module, :derive)}
   end
 
+  defp warn_on_duplicate_struct_key([], _) do
+    nil
+  end
+
+  defp warn_on_duplicate_struct_key([{key, _} | rest], key) do
+    IO.warn(~s(duplicate key #{inspect(key)} found in struct))
+    warn_on_duplicate_struct_key(rest, key)
+  end
+
+  defp warn_on_duplicate_struct_key([{key, _} | rest], _) do
+    warn_on_duplicate_struct_key(rest, key)
+  end
```

Comment:
What about this implementation:

```elixir
defp warn_on_duplicate_struct_key([{key, _} | [{key, _} | _] = rest]) do
  IO.warn(~s(duplicate key #{inspect(key)} found in struct))
  warn_on_duplicate_struct_key(rest)
end

defp warn_on_duplicate_struct_key(_other) do
  :ok
end
```

?

Reply:
```

  defp warn_on_duplicate_struct_key([]) do
    :ok
  end

  defp warn_on_duplicate_struct_key([{key, _} | [{key, _} | _] = rest]) do
    IO.warn("duplicate key #{inspect(key)} found in struct")
    warn_on_duplicate_struct_key(rest)
  end

  defp warn_on_duplicate_struct_key([_ | rest]) do
    warn_on_duplicate_struct_key(rest)
  end
```

Reply:
Oh yes, good catch, but yeah so that we can only have one argument and don't have to start with `nil` as the first key :P)